### ISSUE TITLE
Enabling require_code_owner_reviews also enables dismissal_restrictions

### ISSUE THREAD

**Issue by joebowbeer** [terraform-providers/terraform-provider-github#267]
### Terraform Version
Terraform v0.11.14
+ provider.github v2.2.0

### Affected Resource(s)
- github_branch_protection

### Terraform Configuration Files

```
resource "github_branch_protection" "my-repo" {
  repository     = "my-repo.name"
  branch         = "master"
  enforce_admins = true

  required_pull_request_reviews {
    dismiss_stale_reviews      = true
    require_code_owner_reviews = true
  }
}
```

### Expected Behavior
Requiring code owner reviews should not affect dismissal_restrictions.

### Actual Behavior
dismissal_restrictions is enabled in GitHub console, but no one is listed as being able to dismiss:

```
Restrict who can dismiss pull request reviews
Specify people or teams allowed to dismiss pull request reviews. 
```

If, furthermore, the restrictions apply to admins (enforce_admins), then absolutely no one except the reviewer will be able to dismiss.

If I then disable `dismissal_restrictions` in GitHub and refresh the plan, no changes are detected.
### References
- #50 
- https://github.com/google/go-github/pull/791#issuecomment-518994075

**Comment by verbruggenalex** [comment]
I'm having the same behavior. I think if the option is not defined the checkbox should stay unchecked.

Subscribing to this issue.

**Comment by joebowbeer** [comment]
I verified that `dismissal_restrictions` can be disabled by setting it to an empty object:

```bash
curl -X PUT \
-H "Accept: application/vnd.github.v3+json" \
-H "Authorization: token :token" \
-d '{"dismiss_stale_reviews": true, "required_status_checks": null, "enforce_admins": null, "required_pull_request_reviews": {"dismissal_restrictions": {}}, "restrictions": null}' \
https://api.github.com/repos/:owner/:repo/branches/master/protection
```

Before running that command the dismissal_restrictions created by terraform looked like:

```json
"dismissal_restrictions": {
  "users": [

  ],
  "teams": [

  ]
}
```

After running this command, dismissal_restrictions was no longer present in branch protection response.

Does the current terraform provider do the right thing? It's unclear from looking at the code

**Comment by verbruggenalex** [comment]
Yes this seems to be the correct behavior: https://developer.github.com/v3/repos/branches/#update-branch-protection

> Specify which users and teams can dismiss pull request reviews. Pass an empty dismissal_restrictions object to disable. User and team dismissal_restrictions are only available for organization-owned repositories. Omit this parameter for personal repositories.

Case closed I guess ...

**Comment by verbruggenalex** [comment]
Ah now I see. The `dismissal_restrictions` is not available in the terraform github provider settings. So there's no way to uncheck the checkbox. I think some refactoring is needed to solve this in terraform.

**Comment by joebowbeer** [comment]
Summarizing where I'm at now.

At one time terraform created this erroneous state, which prevents anyone from dismissing a review:

```json
"dismissal_restrictions": {
  "users": [

  ],
  "teams": [

  ]
}
```

I don't know whether terraform continues to produce this state, but I do know that terraform is unable to clean up this state.

I verified that the go-github was broken wrt dismissal_restrictions but is now doing the right thing.

**Comment by thekbb** [comment]
@joebowbeer It's still an issue with github provider `2.4.1`

**Comment by jai** [comment]
It's not just `require_code_owner_reviews` - the way it's structured, any usage of `required_pull_request_reviews` results in empty arrays being sent for `dismissal_users` and `dismissal_teams` which leads to the aforementioned setting on github.

I've dirty-fixed it so it doesn't write that state (https://github.com/jai/terraform-provider-github/commit/784033501f4fce4d72f508a6240714d06ac5ecc4), but still can't find a way to test it. Not sure if the test suite supports offline testing of resource <-> Github request internally. The way the acceptance tests are set up makes initial setup of test scenario seem super tedious.

Also need to fix it so it goes the other way around so that the incorrectly-written state can be removed.

The way `dismissal_users` and `dismissal_teams` is structured doesn't seem right to me given that ultimately it's 1 checkbox with 2 lists below it. Need to learn more about Terraform schema, then will propose a fix (and the corresponding tests).

**Comment by anGie44** [comment]
hi @jai, are you working on this issue? I posted a PR that relates to this just now before reading your latest comment here with more detail. lmk if I can help in any way! Also, I can close my PR in favor of what you plan on working on :)

**Comment by jai** [comment]
I made a tweak that I think fixes this specific issue, I forgot where I put it. Sorry it’s late 😴 I had to take a couple weeks’ break from devop’ing to do some Kotlin but I’m back on it, I’ll check out your PR shortly.

**Comment by anGie44** [comment]
no worries (or rush) @jai ! just wanted to check I wasn't duplicating any work on this. I appreciate the review :)
